src/Raven.Server/Documents/Sharding/Queries/AbstractShardedQueryProcessor.cs

```diff
@@ -126,23 +126,26 @@ public virtual ValueTask InitializeAsync()
 
         if (_queryTemplates == null)
         {
-            RewriteQueryIfNeeded(ref queryTemplate);
+            using var rewrittenQuery = RewriteQueryIfNeeded(queryTemplate);
 
             _queryTemplates = new(RequestHandler.DatabaseContext.ShardCount);
 
+            // need to do this only once because the query is the same for all shards
```

Comment:
Why do we need to materialize that to a string?

Reply:
@lastav5 What we need to think about and check is how much this change affects the managed memory usage and GC.

Reply:
If we write it to the stream as a blittable we would have to clone it first, which was worse in performance when I tried. Also see this: https://github.com/ravendb/ravendb/pull/15525/files

Reply:
Maybe we do not need to clone that blittable at all, since we are only reading from it?

Reply:
@ppekrol Tried that. But `AsyncBlittableJsonTextWriter` isn't thread safe when reading from context

Reply:
Can you separate contexts & writers per shard, but read from the same location, no?

Reply:
@ayende Not sure that I understood you entirely, but If I use the separate context that each shard has I'll have to do `Query.ToJson()` (above) on each anyway, which is more expensive than the `ToString()`.

Reply:
What @ayende is saying is that each shard will have its own Writer with Context, but the source blittable that you read will be on the original context.

Reply:
IIRC it is a problem because of https://github.com/ravendb/ravendb/blob/43fd8838658c1c3d9519f23dcc4f89d1bc86a0d5/src/Sparrow/Json/AbstractBlittableJsonTextWriter.cs#L107 
which is allocating on the original context:
https://github.com/ravendb/ravendb/blob/43fd8838658c1c3d9519f23dcc4f89d1bc86a0d5/src/Sparrow/Json/BlittableJsonReaderObject.cs#L898

Reply:
Each blittable json writer can have its own context. Note that this context is passed from the writer, it's not the same as the blittable context

Reply:
Maybe I missed something, but this context is from the `BlittableJsonReaderObject` itself

Reply:
No, it's the context of the writer.

Reply:
Addressed in https://github.com/ravendb/ravendb/pull/15983